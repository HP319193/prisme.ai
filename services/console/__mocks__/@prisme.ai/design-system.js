export function ListItem({ children = null }) {
  return children;
}

export function SearchInput({ children = null }) {
  return children;
}

export function Input({ children = null }) {
  return children;
}

export function Space({ children = null }) {
  return children;
}

export function Title({ children = null }) {
  return children;
}

export function Button({ children = null }) {
  return children;
}

export function SidePanel({ children = null }) {
  return children;
}

export function Menu({ children = null }) {
  return children;
}

export function Dropdown({ children = null }) {
  return children;
}

export function Divider({ children = null }) {
  return children;
}

export function Avatar({ children = null }) {
  return children;
}

export function Collapse({ children = null }) {
  return children;
}

export function Feed({ children = null }) {
  return children;
}

export function Layout({ children = null }) {
  return children;
}

export function Section({ children = null }) {
  return children;
}

export function CollapseItem({ children = null }) {
  return children;
}

export function PageHeader({ children = null }) {
  return children;
}

export function Row({ children = null }) {
  return children;
}

export function Col({ children = null }) {
  return children;
}

export function MenuTab({ children = null }) {
  return children;
}

export function Table({ children = null }) {
  return children;
}

export function Text({ children = null }) {
  return children;
}

export function TagEditable({ children = null }) {
  return children;
}

export function EditableTitle({ children = null }) {
  return children;
}

export function Loading({ children = null }) {
  return children;
}

export function Modal({ children = null }) {
  return children;
}

export function Card({ children = null }) {
  return children;
}

export function Switch({ children = null }) {
  return children;
}

export function Tooltip({ children = null }) {
  return children;
}

export const notification = {
  error: jest.fn(),
  confirm: jest.fn(),
  info: jest.fn(),
  success: jest.fn(),
};

export function Select({ children = null }) {
  return children;
}

export function Popover({ children = null }) {
  return children;
}

export function LocalizedInput({ children = null }) {
  return children;
}

export function SchemaForm({ children = null }) {
  return children;
}

export function SchemaFormDescription({ children = null }) {
  return children;
}

export const schemaTypes = [
  'string',
  'localized:string',
  'number',
  'localized:number',
  'boolean',
  'localized:boolean',
  'object',
  'array',
];

export const UIWidgetsByType = {
  string: ['upload', 'select'],
};

const localize = (text) =>
  text && typeof text === 'object' ? text.en || '' : text;
const localizeSchemaForm = (schema) => schema;

export function useLocalizedText(t, language) {
  return {
    localize,
    localizeSchemaForm,
  };
}
