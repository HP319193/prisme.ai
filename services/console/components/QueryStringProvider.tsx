import { useRouter } from 'next/router';
import {
  createContext,
  FC,
  useContext,
  useState,
  useEffect,
  Dispatch,
  SetStateAction,
} from 'react';

interface QueryStringProviderContext {
  setQueryString: Dispatch<SetStateAction<URLSearchParams>>;
  queryString: URLSearchParams;
}

export const context = createContext<QueryStringProviderContext>({
  setQueryString() {},
  queryString: new URLSearchParams(),
});

export const useQueryString = () => useContext(context);

const initialQueryString = new URLSearchParams();

export const QueryStringProvider: FC = ({ children }) => {
  const [queryString, setQueryString] = useState(initialQueryString);
  const router = useRouter();

  useEffect(() => {
    setQueryString(new URLSearchParams(window.location.search));
  }, [router]);

  useEffect(() => {
    const newQueryString = new URLSearchParams(window.location.search);
    if (initialQueryString.toString() === newQueryString.toString()) return;
    setQueryString(newQueryString);
  }, []);

  useEffect(() => {
    if (queryString === initialQueryString) return;
    const q = queryString.toString();

    history.pushState(
      null,
      '',
      `${window.location.pathname}${q ? `?${q}` : ''}`
    );
  }, [queryString]);
  return (
    <context.Provider value={{ queryString, setQueryString }}>
      {children}
    </context.Provider>
  );
};

export default QueryStringProvider;
