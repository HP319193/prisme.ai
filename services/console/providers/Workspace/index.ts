export * from './WorkspaceProvider';
export { default } from './WorkspaceProvider';
