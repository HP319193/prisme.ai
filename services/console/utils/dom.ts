export function selectText(element: HTMLElement) {
  const selection = window.getSelection()!;
  const range = document.createRange();

  range.selectNodeContents(element);
  selection.removeAllRanges();
  selection.addRange(range);

  return selection.toString();
}
