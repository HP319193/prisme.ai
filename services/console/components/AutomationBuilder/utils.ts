import { Schema } from '@prisme.ai/design-system';

const HAS_NO_FORM = ['all', 'conditions'];
export const instructionHasForm = (name: string, schema: Schema) => {
  return !HAS_NO_FORM.includes(name) && schema.type;
};
