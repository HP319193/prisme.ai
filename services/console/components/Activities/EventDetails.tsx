import { FC, MouseEvent, useCallback, useMemo } from 'react';
import { Table, Tooltip } from '@prisme.ai/design-system';
import { Event } from '@prisme.ai/sdk';
import { useTranslation } from 'next-i18next';
import { selectText } from '../../utils/dom';
import { truncate } from '../../utils/strings';
import {
  FilterOutlined,
  PlayCircleOutlined,
  PlusCircleOutlined,
} from '@ant-design/icons';
import { useWorkspace } from '../../providers/Workspace';
import { useQueryString } from '../../providers/QueryStringProvider';

interface EventsDetailsProps {
  event: Event<Date>;
}

const PAYLOAD_TRUNCATE_LENGTH = 400;

const truncatePayload = (payloadValue: string) => {
  if (payloadValue && payloadValue.length > PAYLOAD_TRUNCATE_LENGTH) {
    return truncate(payloadValue, PAYLOAD_TRUNCATE_LENGTH);
  }
  return payloadValue;
};

interface EventRecord {
  key: string;
  name: string;
  value: any;
  payloadValue: any;
  fullPayload: string;
}

export const EventDetails: FC<EventsDetailsProps & { replay: () => void }> = ({
  replay,
  event,
}) => {
  const { t } = useTranslation('workspaces');
  const { setQueryString } = useQueryString();
  const dataSource = useMemo(() => {
    const stringifiedPayload = JSON.stringify(event.payload, null, ' ');
    return [
      {
        key: 'type',
        name: 'type',
        value: event.type,
      },
      event.source.appInstanceFullSlug
        ? {
            key: 'source.appInstanceFullSlug',
            name: 'source.appInstanceFullSlug',
            value: event.source.appInstanceFullSlug,
          }
        : false,
      event.source.automationSlug
        ? {
            key: 'source.automationSlug',
            name: 'source.automationSlug',
            value: event.source.automationSlug,
          }
        : false,
      {
        key: 'payload',
        name: 'payload',
        value: (
          <pre>
            <code>{truncatePayload(stringifiedPayload)}</code>
          </pre>
        ),
        payloadValue: event.payload && (
          <pre>
            <code>{stringifiedPayload}</code>
          </pre>
        ),
        fullPayload: stringifiedPayload,
      },
      {
        key: 'source.userId',
        name: 'source.userId',
        value: event.source.userId,
      },
      {
        key: 'source.sessionId',
        name: 'source.sessionId',
        value: event.source.sessionId,
      },
      {
        key: 'source.correlationId',
        name: 'source.correlationId',
        value: event.source.correlationId,
      },
      event.error
        ? {
            key: 'error.error',
            name: 'error.error',
            value: event.error?.error,
          }
        : false,
      event.error
        ? {
            key: 'error.message',
            name: 'error.message',
            value: event.error?.message,
          }
        : false,
      event.error
        ? {
            key: 'error.details',
            name: 'error.details',
            value: event.error?.details && (
              <pre>
                <code>{JSON.stringify(event.error?.details, null, ' ')}</code>
              </pre>
            ),
          }
        : false,
      event.target
        ? {
            key: 'target',
            name: 'target',
            value: event.target && (
              <pre>
                <code>{JSON.stringify(event.target, null, ' ')}</code>
              </pre>
            ),
          }
        : false,
      {
        key: 'id',
        name: 'id',
        value: event.id,
      },
      {
        key: 'size',
        name: 'size',
        value: event.size,
      },
    ].filter(Boolean) as EventRecord[];
  }, [event]);
  const onRowClick = useCallback(({ target }: MouseEvent) => {
    const valueTd = (
      target as HTMLTableRowElement
    ).parentNode?.querySelectorAll('td');
    if (valueTd && valueTd.length === 3) {
      selectText(valueTd[2]);
    }
  }, []);

  const isEmit = event?.source?.serviceTopic === 'topic:runtime:emit';

  const addFilter = useCallback(
    (k: string, v: string) => {
      setQueryString((prev) => {
        const newQuery = new URLSearchParams(prev);
        newQuery.set(k, v);
        return newQuery;
      });
    },
    [setQueryString]
  );

  return (
    <div className="relative">
      <Table
        dataSource={dataSource}
        columns={[
          { title: t('events.details.name'), dataIndex: 'name', key: 'name' },
          {
            title: t('events.details.value'),
            dataIndex: 'value',
            key: 'value',
            render: (children, data) => {
              if (typeof children === 'object') return children;
              return (
                <>
                  {children}
                  <Tooltip title={t('events.details.addFilter')}>
                    <button
                      className="relative text-gray ml-2"
                      onClick={() => addFilter(data.name, data.value)}
                    >
                      <FilterOutlined />
                      <PlusCircleOutlined className="absolute bottom-0 -right-1 text-xs" />
                    </button>
                  </Tooltip>
                </>
              );
            },
          },
        ]}
        bordered
        pagination={{ pageSize: 50, position: [] }}
        scroll={{ y: 500 }}
        expandable={{
          expandedRowRender: (record) => record.payloadValue,
          rowExpandable: (record) =>
            record.name === 'payload' &&
            !!record.fullPayload &&
            record.fullPayload.length > PAYLOAD_TRUNCATE_LENGTH,
        }}
        expandRowByClick
        onRow={() => ({
          onClick: onRowClick,
        })}
      />
      {isEmit && (
        <Tooltip title={t('feed.replay')} placement="right">
          <button
            className="absolute top-[1.3rem] left-[1.3rem]"
            onClick={replay}
          >
            <PlayCircleOutlined />
          </button>
        </Tooltip>
      )}
    </div>
  );
};

export const LinkedEventDetails: FC<Event<Date>> = (event) => {
  const { events } = useWorkspace();
  const replay = useCallback(() => {
    events.emit(event.type, event.payload);
  }, [events, event.payload, event.type]);
  return useMemo(
    () => <EventDetails event={event} replay={replay} />,
    [event, replay]
  );
};

export default LinkedEventDetails;
