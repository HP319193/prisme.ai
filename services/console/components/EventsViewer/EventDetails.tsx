import { FC, useCallback, useMemo } from "react";
import { Event } from "../../api/types";
import { DataTable, DataTableRowClickEventParams } from "primereact/datatable";
import { Column } from "primereact/column";
import { useDateFormat } from "../../utils/dates";
import { useTranslation } from "next-i18next";
import { selectText } from "../../utils/dom";

interface EventsDetailsProps extends Event<Date> {}

export const EventDetails: FC<EventsDetailsProps> = (event) => {
  const { t } = useTranslation("workspaces");
  const formatDate = useDateFormat();
  const value = useMemo(
    () => [
      {
        name: "id",
        value: event.id,
      },
      {
        name: "type",
        value: event.type,
      },
      {
        name: "createdAt",
        value: formatDate(event.createdAt, { format: "yyyy-MM-dd hh:mm" }),
      },
      {
        name: "source.app",
        value: event.source.app,
      },
      {
        name: "source.userId",
        value: event.source.userId,
      },
      {
        name: "source.workspaceId",
        value: event.source.workspaceId,
      },
      {
        name: "source.host.service",
        value: event.source.host.service,
      },
      {
        name: "source.correlationId",
        value: event.source.correlationId,
      },
      {
        name: "payload",
        value: event.payload && (
          <pre>
            <code>{JSON.stringify(event.payload, null, " ")}</code>
          </pre>
        ),
      },
      {
        name: "error.error",
        value: event.error?.error,
      },
      {
        name: "error.message",
        value: event.error?.message,
      },
      {
        name: "error.details",
        value: event.error?.details && (
          <pre>
            <code>{JSON.stringify(event.error?.details, null, " ")}</code>
          </pre>
        ),
      },
    ],
    [event, formatDate]
  );
  const onRowClick = useCallback(
    ({ originalEvent: { target } }: DataTableRowClickEventParams) => {
      const valueTd = (
        target as HTMLTableRowElement
      ).parentNode?.querySelectorAll("td")[1]!;
      selectText(valueTd);
    },
    []
  );

  return (
    <DataTable value={value} onRowClick={onRowClick}>
      <Column field="name" header={t("events.details.name")} />
      <Column field="value" header={t("events.details.value")} />
    </DataTable>
  );
};

export default EventDetails;
