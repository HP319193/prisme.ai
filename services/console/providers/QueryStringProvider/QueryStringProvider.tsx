import { useRouter } from 'next/router';
import {
  createContext,
  FC,
  useState,
  useEffect,
  Dispatch,
  SetStateAction,
} from 'react';
import { useContext } from '../../utils/useContext';

export interface QueryStringProviderContext {
  setQueryString: Dispatch<SetStateAction<URLSearchParams>>;
  queryString: URLSearchParams;
}

export const queryStringContext = createContext<
  QueryStringProviderContext | undefined
>(undefined);

export const useQueryString = () =>
  useContext<QueryStringProviderContext>(queryStringContext);

const initialQueryString = new URLSearchParams();

export const QueryStringProvider: FC = ({ children }) => {
  const [queryString, setQueryString] = useState(initialQueryString);
  const router = useRouter();

  useEffect(() => {
    setQueryString(new URLSearchParams(window.location.search));
  }, [router]);

  useEffect(() => {
    const newQueryString = new URLSearchParams(window.location.search);
    if (initialQueryString.toString() === newQueryString.toString()) return;
    setQueryString(newQueryString);
  }, []);

  useEffect(() => {
    // Clean empty fields
    const q = new URLSearchParams(
      Object.fromEntries(
        Array.from(queryString.entries()).filter(([k, v]) => k && v)
      )
    ).toString();
    if (q === window.location.search) return;

    history.pushState(
      null,
      '',
      `${window.location.pathname}${q ? `?${q}` : ''}`
    );
  }, [queryString]);
  return (
    <queryStringContext.Provider value={{ queryString, setQueryString }}>
      {children}
    </queryStringContext.Provider>
  );
};

export default QueryStringProvider;
