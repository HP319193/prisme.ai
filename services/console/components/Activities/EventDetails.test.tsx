import EventDetails from './EventDetails';
import renderer, { act } from 'react-test-renderer';
import { Event } from '@prisme.ai/sdk';
import { FC } from 'react';
import { selectText } from '../../utils/dom';
import { Table } from '@prisme.ai/design-system';
import { workspaceContext } from '../../providers/Workspace';
import workspaceContextValue from '../../providers/Workspace/workspaceContextValue.mock';
import QueryStringProvider from '../../providers/QueryStringProvider';

jest.mock('@prisme.ai/design-system', () => {
  const Table: FC = ({ children }) => <div className="Table">{children}</div>;
  return {
    Table: Table,
  };
});
jest.mock('../../utils/dom', () => {
  const mock = jest.fn();
  return {
    selectText: mock,
  };
});

jest.mock('../../utils/dates', () => {
  const mock = jest.fn();
  return {
    useDateFormat: () => mock,
  };
});

const event: Event<Date> = {
  id: '1',
  createdAt: new Date('2022-01-01'),
  type: 'foo',
  source: {
    correlationId: '2',
    host: {
      service: 'acme',
    },
  },
  size: 0,
};

it('should render', () => {
  const root = renderer.create(
    <QueryStringProvider>
      <workspaceContext.Provider value={workspaceContextValue}>
        <EventDetails {...event} />
      </workspaceContext.Provider>
    </QueryStringProvider>
  );
  expect(root.toJSON()).toMatchSnapshot();
});

it('should select on row click', () => {
  const root = renderer.create(
    <QueryStringProvider>
      <workspaceContext.Provider value={workspaceContextValue}>
        <EventDetails {...event} />
      </workspaceContext.Provider>
    </QueryStringProvider>
  );

  const tr = document.createElement('tr');
  const td0 = document.createElement('td');
  const td1 = document.createElement('td');
  const td2 = document.createElement('td');
  tr.appendChild(td0);
  tr.appendChild(td1);
  tr.appendChild(td2);
  act(() => {
    root.root.findByType(Table).props.onRow().onClick({
      target: td0,
    });
  });
  expect(selectText).toHaveBeenCalledWith(td2);
});
