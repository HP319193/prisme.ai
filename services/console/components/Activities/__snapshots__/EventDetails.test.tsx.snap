// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div
  className="relative"
>
  <div
    className="Table"
  />
</div>
`;
